wagtail_transfer/locators.py

```diff
@@ -70,7 +77,57 @@ def attach_uid(self, instance, uid):
             uid=uid, defaults={'content_type': self.content_type, 'local_id': instance.pk}
         )
 
+    def uid_from_json(self, json_uid):
+        """
+        Convert the UID representation originating from JSON data into the native type used by
+        this locator
+        """
+        # No conversion necessary, as UID is a string, and JSON handles those fine...
+        return json_uid
+
+
+class FieldLocator:
+    def __init__(self, model, fields):
+        if model._meta.parents:
+            raise ImproperlyConfigured(
+                "FieldLocator cannot be used on MTI subclasses (got %r)" % model
+            )
+        self.model = model
+        self.fields = fields
+
+    def get_uid_for_local_id(self, id):
+        # For field-based lookups, the UID is a tuple of field values
+        return self.model.objects.values_list(*self.fields).get(pk=id)
+
+    def attach_uid(self, instance, uid):
+        # UID is derived directly from the object data, so nothing needs to be done to associate
+        # the UID with the object
+        pass
+
+    def uid_from_json(self, json_uid):
+        # A UID coming from JSON data will arrive as a list (because JSON has no tuple type),
+        # but we need a tuple because the importer logic expects a hashable type that we can use
+        # in sets and dict keys
+        return tuple(json_uid)
+
+    def find(self, uid):
+        # pair up field names with their respective items in the UID tuple, to form a filter dict
+        # that we can use for an ORM lookup
+        filters = dict(zip(self.fields, uid))
+
+        try:
+            return self.model.objects.get(**filters)
+        except self.model.DoesNotExist:
+            return None
+
 
 @lru_cache(maxsize=None)
 def get_locator_for_model(model):
```

Comment:
I really like the field lookup ability here - I think that's really useful. However, I can see people wanting to extend it with a bit of their own logic. Do you think we could implement an approach similar to FieldAdapters here as well, which makes it a bit easier to insert a custom locator? At the moment FieldLocator vs IDMappingLocator is hardcoded.

Reply:
I think our best bet here is to wait and see what sort of use-cases come up. Hopefully adding new locator types _internally_ should be straightforward now - opening it up to external code is a much bigger maintenance burden and something I'd prefer to avoid until there's a clear need for it.

Reply:
Okay, that makes sense.